File: web-app/src/server/rag/cache/semantic-cache.ts
import { createHash } from "node:crypto";
import { prisma } from "@/server/db";
import { vectorQueries } from "@/server/db/vector-queries";
import type { Source } from "@/server/rag/types";
import { CACHE_SIMILARITY_THRESHOLD, CACHE_TTL_SECONDS } from "@/server/rag/types";
import { createLogger } from "@/server/lib/logger";

const logger = createLogger("semantic-cache");

export interface CachePayload {
  answer: string;
  sources: Source[];
}

export interface CachedResponse {
  answer: string;
  sources: Source[];
  retrievalPath: string;
  latencyMs: number;
  isCached: boolean;
}

// Re-exported from vector-queries for backward compatibility with any existing
// test code that imports this type directly from semantic-cache.
export type { CacheSimRow as CacheHitRow } from "@/server/db/vector-queries";

/**
 * Multi-tier semantic cache (WEB_APP_PLAN §7; ported from `src/semantic_cache.py`):
 * - Tier 1: SHA-256 exact string match.
 * - Tier 2: pgvector cosine similarity >= 0.97.
 * TTL (7 days) is enforced at read and write time.
 * - Race-safe writes: INSERT … ON CONFLICT DO UPDATE eliminates the TOCTOU race.
 * - Raw SQL delegated to `vectorQueries` (src/server/db/vector-queries.ts) —
 *   the single source of truth for all pgvector queries.
 */
export class SemanticCache {
  constructor(
    private readonly threshold: number = CACHE_SIMILARITY_THRESHOLD,
    private readonly ttlSeconds: number = CACHE_TTL_SECONDS,
  ) {}

  private hashQuery(query: string): string {
    return createHash("sha256").update(query.trim().toLowerCase()).digest("hex");
  }

  async checkCache(query: string, queryVector: number[] | null): Promise<CachedResponse | null> {
    const qHash = this.hashQuery(query);
    const now = new Date();

    try {
      const exact = await prisma.semanticCacheEntry.findUnique({
        where: { queryHash: qHash },
      });
      if (exact && exact.expiresAt > now) {
        logger.info(`[CACHE HIT - TIER 1 EXACT] query: '${query.slice(0, 40)}'`);
        return this.toCachedResponse(exact, "TIER_1_EXACT_CACHE_HIT", 1.2);
      }

      if (queryVector && queryVector.length > 0) {
        const rows = await vectorQueries.findSimilarCacheEntry(prisma, queryVector, now);

        if (rows.length > 0) {
          const sim = Number(rows[0].sim);
          if (sim >= this.threshold) {
            logger.info(
              `[CACHE HIT - TIER 2 VECTOR] sim ${sim.toFixed(4)} for query: '${query.slice(0, 40)}'`,
            );
            return {
              answer: this.extractAnswer(rows[0].responseJson),
              sources: this.extractSources(rows[0].responseJson),
              retrievalPath: `TIER_2_VECTOR_CACHE_HIT (Sim: ${sim.toFixed(3)})`,
              latencyMs: 12,
              isCached: true,
            };
          }
        }
      }
    } catch (error) {
      logger.warn({ error: String(error) }, "[CACHE] check failed");
    }

    return null;
  }

  async addToCache(
    query: string,
    queryVector: number[],
    responseData: CachePayload,
    parentDocIds: string[] = [],
  ): Promise<void> {
    const qHash = this.hashQuery(query);
    const now = new Date();
    const expiresAt = new Date(now.getTime() + this.ttlSeconds * 1000);
    const payload = JSON.parse(
      JSON.stringify({ answer: responseData.answer, sources: responseData.sources }),
    );

    try {
      /**
       * Delegates to vectorQueries.upsertCacheEntry which issues an atomic
       * INSERT … ON CONFLICT DO UPDATE — the single authoritative vector upsert
       * path (src/server/db/vector-queries.ts).
       */
      await vectorQueries.upsertCacheEntry(prisma, {
        queryHash: qHash,
        queryText: query,
        queryVector,
        responseJson: JSON.stringify(payload),
        parentDocIds,
        now,
        expiresAt,
      });
      logger.info(`[CACHE] upserted entry for query: '${query.slice(0, 40)}'`);
    } catch (error) {
      logger.warn({ error: String(error) }, "[CACHE] write failed");
    }
  }

  async clearAll(): Promise<boolean> {
    try {
      await prisma.semanticCacheEntry.deleteMany({});
      logger.info("[CACHE] cleared all entries");
      return true;
    } catch (error) {
      logger.warn({ error: String(error) }, "[CACHE] clear failed");
      return false;
    }
  }

  async invalidateForDocument(docId: string): Promise<number> {
    try {
      const entries = await prisma.semanticCacheEntry.findMany({
        where: { parentDocIds: { has: docId } },
      });
      const ids = entries.map((entry) => entry.id);
      if (ids.length > 0) {
        await prisma.semanticCacheEntry.deleteMany({ where: { id: { in: ids } } });
      }
      logger.info(`[CACHE] invalidated ${ids.length} entries for document '${docId}'`);
      return ids.length;
    } catch (error) {
      logger.warn({ error: String(error) }, "[CACHE] invalidate failed");
      return 0;
    }
  }

  private toCachedResponse(
    row: {
      responseJson: unknown;
    },
    retrievalPath: string,
    latencyMs: number,
  ): CachedResponse {
    return {
      answer: this.extractAnswer(row.responseJson),
      sources: this.extractSources(row.responseJson),
      retrievalPath,
      latencyMs,
      isCached: true,
    };
  }

  private extractAnswer(responseJson: unknown): string {
    const data = responseJson as { answer?: string };
    return data?.answer ?? "";
  }

  private extractSources(responseJson: unknown): Source[] {
    const data = responseJson as { sources?: Source[] };
    return Array.isArray(data?.sources) ? data.sources : [];
  }
}

export const semanticCache = new SemanticCache();

File: mvp-python/src/semantic_cache.py
"""
Enterprise Multi-Tier Semantic Cache Engine (PostgreSQL + pgvector Backend)
Complies with GUARDRAILS.md 6.4 and CODING_STANDARDS.md Pillar 4.4.
- Tier 0: Negative Cache Rejection for Out-of-Domain queries
- Tier 1: SHA-256 Exact String Match
- Tier 2: 768d Cosine Vector Similarity Match (>= 0.93) via pgvector
"""

import os
import time
import asyncio
import hashlib
import numpy as np
from typing import Optional, List, Dict
from sqlalchemy import select, text, cast, String
from src.database import SessionLocal, SemanticCacheEntry
from src.logging_config import logger

DEFAULT_TTL_SECONDS = 7 * 24 * 3600  # 7 Days TTL
DEFAULT_SIMILARITY_THRESHOLD = 0.93

class SemanticCache:
    """
    Enterprise Multi-Tier Cache Engine (PostgreSQL / pgvector Backend):
    - Tier 1: SHA-256 Exact String Match
    - Tier 2: 768d Cosine Vector Similarity Match (>= 0.93) via pgvector
    - Negative Cache: Instant rejection for out-of-domain terms
    """

    def __init__(self, similarity_threshold: float = DEFAULT_SIMILARITY_THRESHOLD):
        self.threshold = similarity_threshold
        self.negative_terms = set([
            "japan", "stock trading", "algorithmic", "crypto", "recipe", "cooking",
            "nba", "football", "cricket", "python script for trading"
        ])
        
    def _hash_query(self, query: str) -> str:
        return hashlib.sha256(query.strip().lower().encode("utf-8")).hexdigest()

    def check_negative_cache(self, query: str) -> bool:
        q_lower = query.lower()
        return any(term in q_lower for term in self.negative_terms)

    async def check_cache(self, query: str, query_vector: Optional[np.ndarray] = None, bypass_cache: bool = False) -> Optional[dict]:
        if bypass_cache:
            return None

        # Tier 0: Negative Cache Rejection
        if self.check_negative_cache(query):
            logger.info(f"[CACHE HIT - NEGATIVE] Query '{query[:40]}' rejected by Negative Cache.")
            return {
                "answer": "This question falls outside the scope of German immigration, student visas, and university applications.",
                "sources": [],
                "retrieval_path": "NEGATIVE_CACHE_HIT",
                "latency_ms": 0.5,
                "is_cached": True
            }

        q_hash = self._hash_query(query)
        
        try:
            async with SessionLocal() as db:
                # Tier 1: SHA-256 Exact String Match
                result = await db.execute(select(SemanticCacheEntry).filter(SemanticCacheEntry.query_hash == q_hash))
                exact_match = result.scalar_one_or_none()
                if exact_match:
                    logger.info(f"[CACHE HIT - TIER 1 EXACT DB] Match for query: '{query[:40]}'")
                    return {
                        "answer": exact_match.response_json["answer"],
                        "sources": exact_match.response_json.get("sources", []),
                        "retrieval_path": "TIER_1_EXACT_CACHE_HIT",
                        "latency_ms": 1.2,
                        "is_cached": True
                    }

                # Tier 2: pgvector Cosine Similarity Match
                if query_vector is not None:
                    # pgvector cosine distance operator is <=>
                    # similarity = 1 - distance
                    sql = text('''
                        SELECT response_json, 1 - (query_vector <=> CAST(:vec AS vector)) AS sim
                        FROM semantic_cache
                        ORDER BY query_vector <=> CAST(:vec AS vector)
                        LIMIT 1
                    ''')
                    vec_str = "[" + ",".join(str(float(x)) for x in query_vector.tolist()) + "]"
                    result = await db.execute(sql, {"vec": vec_str})
                    match = result.first()
                    if match:
                        sim_score = float(match.sim)
                        if sim_score >= self.threshold:
                            logger.info(f"[CACHE HIT - TIER 2 VECTOR DB] Sim: {sim_score:.4f} >= {self.threshold} for query: '{query[:40]}'")
                            return {
                                "answer": match.response_json["answer"],
                                "sources": match.response_json.get("sources", []),
                                "retrieval_path": f"TIER_2_VECTOR_CACHE_HIT (Sim: {sim_score:.3f})",
                                "latency_ms": 12.0,
                                "is_cached": True
                            }
        except Exception as e:
            logger.warning(f"[WARN] Database Semantic Cache check failed: {e}")
            
        return None

    async def add_to_cache(self, query: str, query_vector: np.ndarray, response_data: dict, parent_doc_ids: List[str] = None, bypass_cache: bool = False):
        if bypass_cache:
            return
        q_hash = self._hash_query(query)
        try:
            async with SessionLocal() as db:
                result = await db.execute(select(SemanticCacheEntry).filter(SemanticCacheEntry.query_hash == q_hash))
                existing = result.scalar_one_or_none()
                if not existing:
                    vec_list = query_vector.tolist() if isinstance(query_vector, np.ndarray) else query_vector
                    new_entry = SemanticCacheEntry(
                        query_hash=q_hash,
                        query_text=query,
                        query_vector=vec_list,
                        response_json=response_data,
                        parent_doc_ids=parent_doc_ids or []
                    )
                    db.add(new_entry)
                    await db.commit()
                    logger.info(f"[SEMANTIC CACHE] Added new entry to PostgreSQL: '{query[:40]}'")
        except Exception as e:
            logger.warning(f"[WARN] Failed to write to PostgreSQL cache: {e}")

    def save_cache(self, query: str, answer_text: str, sources: List[dict] = None, query_vector: Optional[np.ndarray] = None):
        """Synchronous wrapper / alias method for saving query results into cache."""
        response_data = {"answer": answer_text, "sources": sources or []}
        if query_vector is None:
            query_vector = np.zeros(768)
        try:
            loop = asyncio.get_running_loop()
            loop.create_task(self.add_to_cache(query, query_vector, response_data))
        except RuntimeError:
            asyncio.run(self.add_to_cache(query, query_vector, response_data))

    async def clear_all_cache(self) -> bool:
        """Clear all entries from PostgreSQL semantic cache to start completely fresh."""
        try:
            async with SessionLocal() as db:
                await db.execute(text("DELETE FROM semantic_cache"))
                await db.commit()
                logger.info("🧹 [SEMANTIC CACHE] Cleared all entries from PostgreSQL cache!")
                return True
        except Exception as e:
            logger.warning(f"[WARN] Failed to clear PostgreSQL cache: {e}")
            return False

    def clear(self) -> bool:
        """Synchronous alias to clear cache entries (used by CLI / maintenance scripts)."""
        try:
            return asyncio.run(self.clear_all_cache())
        except Exception as e:
            logger.warning(f"[WARN] Cache clear failed: {e}")
            return False

    async def invalidate_cache_for_document(self, doc_id: str) -> int:
        """Invalidate all cache entries derived from a specific document ID."""
        try:
            async with SessionLocal() as db:
                result = await db.execute(
                    select(SemanticCacheEntry).filter(cast(SemanticCacheEntry.parent_doc_ids, String).like(f'%"{doc_id}"%'))
                )
                entries = result.scalars().all()
                count = len(entries)
                for entry in entries:
                    await db.delete(entry)
                await db.commit()
                logger.info(f"[SEMANTIC CACHE] Invalidated {count} cache entries derived from document '{doc_id}'")
                return count
        except Exception as e:
            logger.warning(f"[WARN] Failed to invalidate cache for document '{doc_id}': {e}")
            return 0

_semantic_cache_instance = None

def get_semantic_cache() -> SemanticCache:
    global _semantic_cache_instance
    if _semantic_cache_instance is None:
        _semantic_cache_instance = SemanticCache()
    return _semantic_cache_instance
